base/mysqlxx/Pool.cpp

```diff
@@ -150,14 +151,22 @@ Pool::Entry Pool::tryGet()
 
     initialize();
 
-    /// Searching for connection which was established but wasn't used.
-    for (auto & connection : connections)
+    /// Try to pick an idle connection from already allocated
+    for (auto connection_it = connections.begin(); connection_it != connections.end();)
     {
-        if (connection->ref_count == 0)
+        /// Fixme: Race condition is here b/c this is not synchronized with Pool::Entry's copy-assignment operator
+        if ((*connection_it)->ref_count == 0)
         {
-            Entry res(connection, this);
-            return res.tryForceConnected() ? res : Entry();
+            Entry res(*connection_it, this);
+            if (res.tryForceConnected())  /// Will try to reestablish connection as well
+                return res;
+
+            /// This one is disconnected for good and so must be removed
+            ::delete *connection_it;  /// TODO: Manual memory management is awkward (matches allocConnection() method)
```

Comment:
So, the fix is in these two lines?

Reply:
Basically yes, it is important to get rid of such unrecoverable connections.